src/Clerk.ts

```diff
@@ -276,6 +285,50 @@ export default class Clerk {
     });
   }
 
+  withProtectedRouteSSR(destination: string){
+    return (handler?: (ctxWithSession: SSRPropsContext) => Promise<GetServerSidePropsResult<any>>) => async (ctx: GetServerSidePropsContext) => {
+
+      // GetServerSidePropsContext req does not contain the query.
+      // For the sake of keeping the abstraction cleaner we inject it from ctx
+      // @ts-ignore
+      ctx.req.query = ctx.query;
+
+      try{
+        await this._runMiddleware(ctx.req, ctx.res, this.expressWithSession({ onError: this.strictOnError }));        
+      }catch (err) {
+        return {
+          redirect: {
+            destination,
+            permanent: false,
+          },
+        }
+      }
+
+      return handler ? (await handler(ctx as SSRPropsContext)) : {props: {}};
+    }
+  }
+
+  withSessionSSR(handler: (ctxWithSession: SSRPropsContext) => Promise<GetServerSidePropsResult<any>>, { onError }: MiddlewareOptions = { onError: this.defaultOnError }) {
```

Comment:
❓ If I get it correctly the main difference between withSessionSSR and withSession is the `ctx.req.query = .ctx.query`. Having said that, wouldn't it make sense to merge these middlewares into one and have one withSession handler for SSR and API protected routes?

Reply:
This is one of the main concerns I had when designing the solution. So here are my thoughts.

First of all if we **require** that the API is the same, we can do it. The reasons for not doing that lie with the TypeScript alchemy it would be required to make it work as TypeScript users expect it to.

Couple of my thoughts:

1. We can always say that depending on the usage, the user has to add the typings himself as we do in `requireSession` and `withSession`. But it is kind of as error prone as importing the wrong method + one more step e.g.
![image](https://user-images.githubusercontent.com/15251081/131509727-82f80166-930e-47fc-ac02-9de23710bd35.png)
^^ Here we "trick" TS by giving the user another mishap chance.
E.g. 
```ts
// Works even if it is now req.?session vs req.session
async function handler(
  req: WithSessionProp<NextApiRequest>,
  res: NextApiResponse
) {
 // Here 
});

export default requireSession(handler);
```

2. In the case of the user importing the wrong method, the handler now is already typed to only conform with the signature of the `getServerSideProps` handlers.
E.g. 
```ts
// Misuse of adding withUserSessionSSR to an API route.
export withSessionSSR((req, res) => {
  const query = req.query; // TypeScript will complain
  res.send(404); // TypeScript will complain
});

// Misuse of adding withSession on getServerSideProps
// It does not compile
export const getServerSideProps: GetServerSideProps = withSession(
  async function ({ req }) {
    // Get the user's session based on the request
    const clerkSession = req.session;

    if (!clerkSession) {
      return {
        redirect: {
          destination: "/sign-up",
          permanent: false,
        },
      };
    }

    return { props: {} };
  }
);
```

I am really open to try and consolidate, I am just worried that we will have more issues with TS biting back and hindering us from enhancing appropriately the SSR handlers if we need to later on(_e.g. adding hydration params_).